README.rst
VERACore
========================================

VERACore is a Python application for visualization and engineering analyses of output data from VERA (Virtual Environment for Reactor Applications).
Implemented in Python, it provides instantaneous 2D and 3D images, 1D plots, and alphanumeric data from VERA multi-physics simulations.

For an interactive online example, see `here <https://pvw.kitware.com/vera/>`_.

Installing
----------------------------------------

VERACore can be installed via pip and can run locally or deployed as a service.
To use it locally within a virtual environment, you can run the following commands:

.. code-block:: console

    python3 -m venv ./vera-env
    source ./vera-env/bin/activate
    pip install -U pip
    pip install vera-core

    vera-core --data <path-to-vera-out-file>


Development
----------------------------------------

Setup a virtual environment with a modern Python 3.

.. code-block:: console

    python3 -m venv ./vera-env
    source ./vera-env/bin/activate
    pip install -U pip

Get the code base and build its widgets

.. code-block:: console

    git clone git@github.com:Kitware/VERACore.git
    cd VERACore

    # Build the custom UI widgets
    cd vue-components
    npm i
    npm run build
    cd ..

    # Install local repository into your venv
    pip install -e .

Run the application to test out your changes


.. code-block:: console

    vera-core --server

Licensing
----------------------------------------

VERACore is provided as an open source solution and follows the `Apache Software License <https://github.com/Kitware/VERACore/blob/master/LICENSE>`_


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.3.1 (2023-10-24)

### Fix

* **trame:** Update to trame v3 ([`3802a3e`](https://github.com/Kitware/VERACore/commit/3802a3e875b95559a980b1f9bfb57c083047ad40))

## v1.3.0 (2023-07-17)

### Feature

* **vera-auto-datasets:** Added automatic dataset searching for full-core and scalar datasets, and updated color normalization to be state-wise ([`161cf31`](https://github.com/Kitware/VERACore/commit/161cf317d29e2ec9a881152511d17cd288384bf3))

### Fix

* **opacity:** Changed the volume rendering ([`69381fe`](https://github.com/Kitware/VERACore/commit/69381fec621c2167bb4d3834fa130fb86a98487e))
* **vera-auto-datasets:** Fixed formatting ([`24f8324`](https://github.com/Kitware/VERACore/commit/24f83241f0b1c494691445705ee7206698e7c1a9))

## v1.2.6 (2023-07-13)

### Fix

* Trigger release for new format ([`424c729`](https://github.com/Kitware/VERACore/commit/424c729f50f39723d3465ec96fb6674044fbe73d))

## v1.2.5 (2023-01-11)
### Fix
* **mouse:** Show progress cursor on most widgets ([`eed6859`](https://github.com/Kitware/VERACore/commit/eed6859b80c937881d69ceb2ffca575dd6731822))

## v1.2.4 (2023-01-10)
### Fix
* **busy:** Add a busy indicator ([`8c528ae`](https://github.com/Kitware/VERACore/commit/8c528aeb3e9d37659121b9cf4a446cfad20a6ac1))

## v1.2.3 (2022-12-08)
### Fix
* **style:** Run black ([`6b8cb29`](https://github.com/Kitware/VERACore/commit/6b8cb29ab17f19633efcb09ee9d3e215cd4de02d))
* **html:** Remove root scrollbar ([`2c590c0`](https://github.com/Kitware/VERACore/commit/2c590c0f5415913fbabdb1b6c1a454efa519bd9e))
* **layout:** Remove selection and add margin ([`73c3687`](https://github.com/Kitware/VERACore/commit/73c3687c4b400203eb4e3e0e69b26c12a2845278))
* **volume_view:** Remove bottom scrollbar ([`4af09f3`](https://github.com/Kitware/VERACore/commit/4af09f35c722e1496484c6b00a17977c22ff8f59))

## v1.2.2 (2022-12-07)
### Fix
* **jupyter:** Remove engine ([`3976de3`](https://github.com/Kitware/VERACore/commit/3976de3d7cc437666cd59e1609ca3e5934c75ceb))

## v1.2.1 (2022-12-05)
### Fix
* **options:** Update grid item when option changes ([`d8e989b`](https://github.com/Kitware/VERACore/commit/d8e989b198051ccc25916ff99e7a6d037ffbcd5f))

## v1.2.0 (2022-12-05)
### Feature
* **deploy:** Add initial docker deployment ([`f7695d0`](https://github.com/Kitware/VERACore/commit/f7695d06fe7b1fa16abbc9640773fdf2f6480533))

## v1.1.1 (2022-11-25)
### Fix
* **data:** Allow data to be specified by env variable ([`15da0b4`](https://github.com/Kitware/VERACore/commit/15da0b41068e2907c8e56bef7c68b93fd4c66eb5))

## v1.1.0 (2022-10-07)
### Feature
* **table view:** Add new table view ([`2f13063`](https://github.com/Kitware/VERACore/commit/2f13063d2fdbd91d8cff18c686b0ba5e55e35d6e))
* **volume_view:** Add camera reset button ([`8ed4484`](https://github.com/Kitware/VERACore/commit/8ed4484f6bfd7ade90363d4262c744f64e8e6c5e))
* **selected_line:** Plot selected lines in plots ([`05ea68f`](https://github.com/Kitware/VERACore/commit/05ea68fc5652a32591bd521080fc68825477cda7))
* **volume_view:** Add a volume viewer ([`e280852`](https://github.com/Kitware/VERACore/commit/e280852180f098c38ed363c997b09dfa5b42cd8a))
* **pin_volumes:** Add ability to show pin volumes ([`20d267c`](https://github.com/Kitware/VERACore/commit/20d267cd982da274106eafdbc69ef96a22c2e8b5))
* **core_view:** Add selection ([`4ed5eed`](https://github.com/Kitware/VERACore/commit/4ed5eedf68a86ee0873b38116960181d51c3bf1f))
* **axial_plot:** Added axial plot ([`be52c68`](https://github.com/Kitware/VERACore/commit/be52c6840997871f07201a9df2be59da7cc0734e))
* **assembly_view:** Add assembly view ([`acfb576`](https://github.com/Kitware/VERACore/commit/acfb5769786f322f177125431ff7fd612dc22f76))
* **coreview:** Add core view ([`46d2732`](https://github.com/Kitware/VERACore/commit/46d273200c47ee8567516000f7f24bd07cfd463c))
* **Axial View:** Add axial view and many updates ([`ba843d6`](https://github.com/Kitware/VERACore/commit/ba843d6802359fb55c48d9573ed579a82539ed05))
* **ui:** Add power over time plots with plotly ([`661c25a`](https://github.com/Kitware/VERACore/commit/661c25ace4e91b8fbdcc2bddd5a3578c714739b3))
* **VeraOutFile:** Add VeraOutFile class ([`f14e871`](https://github.com/Kitware/VERACore/commit/f14e871c08f4d3a79239d82b9480dbac150fd091))

### Fix
* **README.rst:** Add `--data <path>` to instructions ([`4f6edb1`](https://github.com/Kitware/VERACore/commit/4f6edb19e8f15801f7bbf10f9068e48d4b2009f3))
* **README.rst:** Fix hyperlink for long description ([`df920b5`](https://github.com/Kitware/VERACore/commit/df920b58fd8776e2bc85406253fadb94854b8e4e))
* **MANIFEST.in:** Add license and select js files ([`ffe0c71`](https://github.com/Kitware/VERACore/commit/ffe0c71d2d19a1853be9b792255f99e640b6a7e7))
* **y axial view:** Fix selected index ([`63e5371`](https://github.com/Kitware/VERACore/commit/63e5371d6d34975ddd1c342556232fe01a323396))
* **double updates:** Only update once after view selections ([`408b64e`](https://github.com/Kitware/VERACore/commit/408b64ece4e6b70ea21b62bd0f39488e059c43e1))
* **tests:** Update import name in pytest ([`612b86b`](https://github.com/Kitware/VERACore/commit/612b86bd88769b25528e36e25a69ff0ade7afdb7))
* **volume:** Small viewer edit ([`5cbcaa2`](https://github.com/Kitware/VERACore/commit/5cbcaa2b7656a1da1c8ddcf58285ae7d7c7162fe))
* **assets:** Add assets to MANIFEST.in ([`26e398f`](https://github.com/Kitware/VERACore/commit/26e398fc3fa445c744fe1bc9a90e9c56c3c845f1))
* **y_axial:** Fix y axial view ([`62e837b`](https://github.com/Kitware/VERACore/commit/62e837b76368c65a96bda2050329ba359b6690ad))
* **color:** Add color map editor ([`ddaf5c8`](https://github.com/Kitware/VERACore/commit/ddaf5c85384d9d36f6eecbd54d7721c2cb2222ee))
* **color_range:** Update color range on array change ([`24ca422`](https://github.com/Kitware/VERACore/commit/24ca422913256a9cd8f5e2d41bc52845142ec8c1))

vera_core/app/ui/__init__.py
import numpy as np

from trame.ui.vuetify import SinglePageLayout
from trame.widgets import client, grid, html, vuetify
from vera_core.widgets import vera
from . import (
    assembly_view,
    axial_plot,
    core_view,
    empty,
    table_view,
    time_plot,
    x_axial_view,
    y_axial_view,
    volume_view,
    assets,
)

DEFAULT_NB_ROWS = 8


def get_next_y_from_layout(layout):
    next_y = 0
    for item in layout:
        y, h = item.get("y", 0), item.get("h", 1)
        if y + h > next_y:
            next_y = y + h
    return next_y


def initialize(server, vera_out_file):
    state, ctrl = server.state, server.controller
    state.trame__title = "VERACore"

    state.setdefault("grid_item_dirty_key", 0)

    # FIXME: For our example, fix this to match VeraView.
    # Come up with a way to autogenerate it.
    state.color_range = (0.0273, 1.95)
    state.selected_layer = 24
    state.selected_assembly = 36
    state.selected_time = 0
    # FIXME ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

    @state.change("selected_time")
    def selected_time_changed(selected_time, **kwargs):
        selected_time = int(selected_time)
        vera_out_file.active_state_index = selected_time

        ctrl.on_vera_out_active_state_index_changed(
            selected_time=selected_time, **kwargs
        )
        # Automatically normalize color scale to current state
        selected_array_changed(state.selected_array)

    @state.change("selected_array")
    def selected_array_changed(selected_array, **kwargs):
        array = vera_out_file.array(selected_array)
        state.color_range = (np.nanmin(array), np.nanmax(array))

    # Keep selected_assembly and selected_assembly_ij in sync
    @state.change("selected_assembly_ij")
    def selected_assembly_ij_changed(selected_assembly_ij, **kwargs):
        i, j = selected_assembly_ij["i"], selected_assembly_ij["j"]
        state.selected_assembly = vera_out_file.core.reduced_core_map_assembly(i, j)

    @state.change("selected_assembly")
    def selected_assembly_changed(selected_assembly, **kwargs):
        i, j = vera_out_file.core.reduced_core_map_ij(selected_assembly)
        state.selected_assembly_ij = dict(i=i, j=j)

    # Go ahead and make sure they are in sync
    selected_assembly_changed(state.selected_assembly)

    # Initialize all visualizations
    state.setdefault("grid_options", [])
    state.setdefault("grid_layout", [])
    assembly_view.initialize(server, vera_out_file)
    axial_plot.initialize(server, vera_out_file)
    x_axial_view.initialize(server, vera_out_file)
    y_axial_view.initialize(server, vera_out_file)
    core_view.initialize(server, vera_out_file)
    table_view.initialize(server, vera_out_file)
    time_plot.initialize(server, vera_out_file)
    volume_view.initialize(server, vera_out_file)
    empty.initialize(server)

    # Reserve the various views
    available_view_ids = [f"{v+1}" for v in range(10)]
    for view_id in available_view_ids:
        state[f"grid_view_{view_id}"] = empty.OPTION

    # X Axial view
    view_id = available_view_ids.pop(0)
    state.grid_layout.append(
        dict(x=0, y=0, w=3, h=17, i=view_id),
    )
    state[f"grid_view_{view_id}"] = x_axial_view.OPTION

    # Y Axial view
    # view_id = available_view_ids.pop(0)
    # state.grid_layout.append(
    #     dict(x=3, y=0, w=3, h=17, i=view_id),
    # )
    # state[f"grid_view_{view_id}"] = y_axial_view.OPTION

    # Core view
    view_id = available_view_ids.pop(0)
    state.grid_layout.append(
        dict(x=6, y=0, w=3, h=9, i=view_id),
    )
    state[f"grid_view_{view_id}"] = core_view.OPTION

    # Assembly view
    view_id = available_view_ids.pop(0)
    state.grid_layout.append(
        dict(x=9, y=0, w=3, h=9, i=view_id),
    )
    state[f"grid_view_{view_id}"] = assembly_view.OPTION

    # Axial plot
    view_id = available_view_ids.pop(0)
    state.grid_layout.append(
        dict(x=6, y=9, w=3, h=8, i=view_id),
    )
    state[f"grid_view_{view_id}"] = axial_plot.OPTION

    # Time plot
    view_id = available_view_ids.pop(0)
    state.grid_layout.append(
        dict(x=9, y=9, w=3, h=8, i=view_id),
    )
    state[f"grid_view_{view_id}"] = time_plot.OPTION

    # Volume view
    view_id = available_view_ids.pop(0)
    state.grid_layout.append(
        dict(x=3, y=0, w=3, h=17, i=view_id),
    )
    state[f"grid_view_{view_id}"] = volume_view.OPTION

    # Table view
    view_id = available_view_ids.pop(0)
    state.grid_layout.append(
        dict(x=0, y=17, w=6, h=10, i=view_id),
    )
    state[f"grid_view_{view_id}"] = table_view.OPTION

    @ctrl.set("grid_add_view")
    def add_view():
        next_view_id = available_view_ids.pop()
        next_y = get_next_y_from_layout(state.grid_layout)
        state.grid_layout.append(
            dict(x=0, w=12, h=DEFAULT_NB_ROWS, y=next_y, i=next_view_id)
        )
        state.dirty("grid_layout")

    @ctrl.set("grid_remove_view")
    def remove_view(view_id):
        available_view_ids.append(view_id)
        state.grid_layout = list(
            filter(lambda item: item.get("i") != view_id, state.grid_layout)
        )

    # Setup main layout
    with SinglePageLayout(server) as layout:
        layout.root.classes = ("{ busy: trame__busy }",)

        # Toolbar
        with layout.toolbar as toolbar:
            toolbar.clear()

            toolbar.height = 36

            html.Img(src=assets.LOGO, height=25)
            vuetify.VSpacer()

            vera.ColorMapEditor(
                v_model="color_range",
                color_preset="jet",
            )

            vuetify.VSpacer()

            with html.Div(
                style="width: 25px",
                classes="mr-2",
            ):
                vuetify.VProgressCircular(
                    indeterminate=True,
                    v_show=("trame__busy",),
                    style="background-color: lightgray; border-radius: 50%",
                    background_opacity=1,
                    bg_color="#01549b",
                    color="#04a94d",
                    size=16,
                    width=3,
                )

            vuetify.VSelect(
                v_model=("selected_array", "pin_powers"),
                items=(
                    "available_arrays",
                    [
                        dict(text=key.replace("_", " ").title(), value=key)
                        for key in vera_out_file.active_state.full_core_datasets.keys()
                    ],
                ),
                hide_details=True,
                dense=True,
                style="max-width: 220px",
            )

            with vuetify.VBtn(icon=True, click=ctrl.grid_add_view):
                vuetify.VIcon("mdi-plus")

        # Main content
        with layout.content:
            layout.content.style = "overflow: auto; margin: 36px 0px 35px; padding: 0;"
            with vuetify.VContainer(
                fluid=True,
                classes="pa-0 fill-height",
                style="user-select: none;",
            ):
                with grid.GridLayout(
                    layout=("grid_layout", []),
                    row_height=30,
                    vertical_compact=True,
                    style="width: 100%; height: 100%;",
                ):
                    with grid.GridItem(
                        v_for="item in grid_layout",
                        key="item.i",
                        v_bind="item",
                        style="touch-action: none;",
                        drag_ignore_from=".drag_ignore",
                    ):
                        with vuetify.VCard(
                            style="height: 100%;",
                            key="grid_item_dirty_key",
                        ):
                            with vuetify.VCardTitle(classes="py-1 px-1"):
                                with vuetify.VMenu(offset_y=True):
                                    with vuetify.Template(
                                        v_slot_activator="{ on, attrs }"
                                    ):
                                        with vuetify.VBtn(
                                            icon=True,
                                            small=True,
                                            v_bind="attrs",
                                            v_on="on",
                                        ):
                                            vuetify.VIcon(
                                                v_text="get(`grid_view_${item.i}`).icon"
                                            )
                                        html.Div(
                                            "{{ get(`grid_view_${item.i}`).label }}",
                                            classes="ml-1 text-subtitle-2",
                                        )
                                    with vuetify.VList(dense=True):
                                        with vuetify.VListItem(
                                            v_for="(option, index) in grid_options",
                                            key="index",
                                            click="""
                                                set(`grid_view_${item.i}`, option);
                                                grid_item_dirty_key++;
                                            """,
                                        ):
                                            with vuetify.VListItemIcon():
                                                vuetify.VIcon(v_text="option.icon")
                                            vuetify.VListItemTitle("{{ option.label }}")
                                vuetify.VSpacer()
                                with vuetify.VBtn(
                                    icon=True,
                                    x_small=True,
                                    click=(ctrl.grid_remove_view, "[item.i]"),
                                ):
                                    vuetify.VIcon(
                                        "mdi-delete-forever-outline", small=True
                                    )
                            vuetify.VDivider()

                            style = "; ".join(
                                [
                                    "position: relative",
                                    "height: calc(100% - 37px)",
                                    "overflow: auto",
                                ]
                            )
                            with vuetify.VCardText(style=style, classes="drag_ignore"):
                                # Add template for value of get(`grid_view_${item.i}`)
                                client.ServerTemplate(
                                    name=("get(`grid_view_${item.i}`).name",)
                                )

        # Footer
        with layout.footer as footer:
            footer.clear()

            footer.height = 35
            with vuetify.VBtn(
                icon=True,
                small=True,
                disabled=("selected_time == 0",),
                click="selected_time--",
            ):
                vuetify.VIcon("mdi-minus")
            html.Div(
                "State {{ selected_time }}",
                classes="text-center",
                style="width: 100px;",
            )
            with vuetify.VBtn(
                icon=True,
                small=True,
                disabled=(f"selected_time == {len(vera_out_file.states) - 1}",),
                click="selected_time++",
            ):
                vuetify.VIcon("mdi-plus")
            vuetify.VDivider(vertical=True, classes="mx-2")
            vuetify.VSlider(
                v_model=("selected_time",),
                min=0,
                max=len(vera_out_file.states) - 1,
                dense=True,
                hide_details=True,
                ticks="always",
                tick_size="4",
                height=35,
                # style="position: relative; top: -10px;",
                # tick_labels=(f"[{','.join([str(v) for v in range(len(vera_out_file.states))])}]", ),
            )
